# MATH 435 Condition Numbers of Hilbert Matrices
James Della-Giustina

# Imports

In [ ]:
import numpy as np
from scipy import linalg
np.set_printoptions(precision=3) # you can adjust this if you feel like it

# Overview (you have to read this)
You should also execute all code cells you encounter along the way

A Hilbert matrix of size $n$ is the $n\times n$ symmetric matrix $H_n = (h_{ij})$ whose $(i,j)$ entry is
\begin{equation*}
    h_{ij} = \frac{1}{i + j - 1} \qquad \text{ for } (i,j)\in \{1, \ldots, n\}\times\{1, \ldots, n\}.
\end{equation*}
Modify the entries in the following matrix to get $H_4$. Write the entries as fractions (not as decimals).
\begin{equation*}
    H_4
    = \begin{bmatrix}
    * & * & * & * \\
    * & * & * & * \\
    * & * & * & * \\
    * & * & * & *
    \end{bmatrix}
\end{equation*}

The Hilbert Matrix $H_4$ is given by:
\begin{equation*}
    H_4
    = \begin{bmatrix}
    1 & \frac{1}{2} & \frac{1}{3} & \frac{1}{4} \\
    \frac{1}{2} & \frac{1}{3} & \frac{1}{4} & \frac{1}{5} \\
    \frac{1}{3} & \frac{1}{4} & \frac{1}{5} & \frac{1}{6} \\
    \frac{1}{4} & \frac{1}{5} & \frac{1}{6} & \frac{1}{7}
    \end{bmatrix}
\end{equation*}

The Hilbert matrices are examples of matrices that are ill-conditioned and the conditioning becomes worse as $n$ becomes larger. We will briefly explore this fact in this assignment.

Python's scipy library has built-in capabilities for generating Hilbert matrices. For example, since we have imported the `linalg` library from `scipy` above, we can simply type `linalg.hilbert(3)` to generate the $3\times 3$ Hilbert matrix:

In [ ]:
H = linalg.hilbert(3)
H

array([[1.   , 0.5  , 0.333],
       [0.5  , 0.333, 0.25 ],
       [0.333, 0.25 , 0.2  ]])

Since we will be exploring the conditioning, let me introduce you to numpy's condition number estimation function `np.linalg.cond()`. For example to get an estimate of the condition number of `H`, we simply type `np.linalg.cond(H)` as follows:

In [ ]:
kappa = np.linalg.cond(H)
kappa

524.0567775860627

As we can see, even when $n = 3$ (which is small in the world of matrix computations), $H_3$ is already suspicious. Let us explore the consequences of this fact in terms of solving equations of the form $H_3 x = b$. Let $z = (1, 1, 1)^\top$ and define $b$ by $b = H_3 z$.  

In [ ]:
z = np.ones(3)
b = H@z

If we numerically compute the solution $x$ to $H_3 x = b$ we would hope to get $x = z$, or at least $x\approx z$. We can numerically compute $x$ by using the `linalg.solve` method as follows:

In [ ]:
x = linalg.solve(H, b)
x

array([1., 1., 1.])

In this case the matrix is small enough (only $3\times 3$) that the `linalg.solve` method successfully produces $x = z$ and this can be seen by visual inspection of $x$. If the matrix were larger we may not be able to tell by visual inspection whether $x\approx z$ unless we were willing to deal with tediously comparing each entry of $x$ to the corresponding entry of $z$. A method for comparing $x$ to $z$ that doesn't rely on visual inspection is to compute the relative error (in this case using the standard Euclidean norm):
\begin{equation*}
    \frac{\|x - z\|}{\|z\|}
\end{equation*}
This can be accomplished using the `linalg.norm` method as follows:

In [ ]:
rel_err = linalg.norm(x - z)/linalg.norm(z)
rel_err

8.022593772267726e-15

In this instance, the fact that $\kappa(H_3) \approx 500$ was not enough to ruin our computation of the solution of $H_3 x = b$.

# Exercise 1

Write a Python function called `solve_hilb` that takes as input, an integer $n$ (the size of a Hilbert matrix) and does the following:
* Creates the $n\times n$ Hilbert matrix $H_n$ and constructs the nonhomogenity $b = H_n z$, where $z = (1, \ldots , 1)^\top\in \mathbb R^n$.
* Solves the linear system $H_n x = b$ for $x$ (you may use built in solver such as `linalg.solve`) then computes the relative error in the computed solution $x$ (relative to the true solution $z$). Use the standard Euclidean norm for the relative error computation.
* returns both the condition number of $H_n$ and the relative error.

All of these tasks were done in the overview in the case $n = 3$, I'm just asking you to automate the process so that one can input $n$ and immediately get the corresponding condition number and relative error.

I've started a skeleton of a function for you in the following cell.

In [ ]:
def solve_hilb(n):
    """
    Solve the equation Hx = b for x, where H is the n-by-n Hilbert matrix (the same n
    as the input n) and b = Hz for z = (1, ..., 1)^t. Return both the conditon number
    of H and the relative error in the solution. 1
    """
    H = linalg.hilbert(n)
    cond_H = np.linalg.cond(H)
    z = np.ones(n)
    b = H@z
    x = linalg.solve(H, b)
    rel_err = linalg.norm(x - z)/linalg.norm(z)

    return cond_H, rel_err

# Exercise 2

Run your `solve_hilb` function on $n = 3$ to be sure you get the same results obtained in the overview.

In [ ]:
# Execute this cell.
solve_hilb(3)

(524.0567775860627, 8.022593772267726e-15)

# Exercise 3

By doing a bit of exploring with feeding different values of $n$ to `solve_hilb`, find the minimum value of $n$ for which the relative error in the computed solution $x$ (relative to the true solution $z = (1, \ldots, 1)^\top$ is larger than or equal to $0.25$. For this $n$, what is the condition number of $H_n$?

*Note: When $n$ gets large enough that the condition number of $H_n$ is intolerably large, Python will raise a warning making you aware of this fact.*

In [ ]:
k=20;
i=1
while i < k:
    print(solve_hilb(i))
    i=i+1


(1.0, 0.0)
(19.281470067903967, 5.661048867003676e-16)
(524.0567775860627, 8.022593772267726e-15)
(15513.738738929662, 4.137409622430382e-14)
(476607.2502419595, 1.6828426299227195e-12)
(14951058.641879061, 1.4242437208427487e-10)
(475367356.9028203, 7.637452450980383e-09)
(15257575505.048504, 6.124089555723088e-08)
(493153563504.81415, 3.8751634185032475e-06)
(16024905787698.326, 8.670390237096913e-05)
(522518134089746.8, 0.000838328777627572)
(1.643080577584755e+16, 0.32491292968691676)
(5.729989400507761e+18, 3.648729201208787)
(3.25920065289962e+17, 3.673009823340255)
(4.4193794499857517e+17, 4.930030954879194)
(8.260202121289064e+18, 3.7578055513989828)
(1.1963372192370002e+18, 6.6573220117364835)
(1.711031621048699e+18, 11.745256954756888)
(9.586490373784824e+18, 5.720717752417546)


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:11: LinAlgWarning: Ill-conditioned matrix (rcond=2.34967e-17): result may not be accurate.
  # This is added back by InteractiveShellApp.init_path()
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:11: LinAlgWarning: Ill-conditioned matrix (rcond=1.28868e-18): result may not be accurate.
  # This is added back by InteractiveShellApp.init_path()
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:11: LinAlgWarning: Ill-conditioned matrix (rcond=1.01671e-18): result may not be accurate.
  # This is added back by InteractiveShellApp.init_path()
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:11: LinAlgWarning: Ill-conditioned matrix (rcond=8.18724e-19): result may not be accurate.
  # This is added back by InteractiveShellApp.init_path()
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:11: LinAlgWarning: Ill-conditioned matrix (rcond=1.1839e-19): result may not be accurate.
  # This 

In [ ]:
solve_hilb(12)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:11: LinAlgWarning: Ill-conditioned matrix (rcond=2.34967e-17): result may not be accurate.
  # This is added back by InteractiveShellApp.init_path()


(1.643080577584755e+16, 0.32491292968691676)

For $n=12$, the Hilbert matrix $H_{12}$ returns an relative error of $\approx .325$ with a condition number $\approx 1.643$. Python accordingly throws the warning of an ill-conditioned matrix at exactly this point.